[TSK-883] [TSK-884] ui-kit: CoachingSessionCalendar: base calendar; student outside a course (variant #1); student in the context of a course (variant #2) #130

ObitoDevv opened the pull request:
Components : 'Calendar' , 'Schedule-session' 

1. Calendar -[In figma](https://www.figma.com/design/8KEwRuOoD5IgxTtFAtLlyS/Just_Do_Ad-1.2?node-id=5210-30504&t=VII2zZycq4mOnDlM-4)
![Image](https://github.com/user-attachments/assets/0cdf1f93-5802-4276-9e92-8634bd4b3327)



2. Schedule-session - [In figma](https://www.figma.com/design/8KEwRuOoD5IgxTtFAtLlyS/Just_Do_Ad-1.2?node-id=5210-29812&m=dev)
![image](https://github.com/user-attachments/assets/f606cff5-7655-4ea7-b254-109076ce7391)
 
 - For isError state I will use the `banner` component and that compo is alreday done in `pre-assessment` pr. 
 - For every `button` I am using callbacks and I will implement the discard button to close the box in the calendar only.
 

Fix #130

ObitoDevv commented:
Added `CoachingSessionCalendar` component with enhanced scheduling functionality. Key details:
- **Variants**:
  - `dragAndDrop`: Enables dragging coaching sessions from `AvailableCoachingSessions` to the calendar, triggers `ScheduleSession` modal for confirmation, and supports event rescheduling via drag-and-drop.
  - `clickToSchedule`: Allows clicking on calendar slots or coach availability to schedule sessions, includes a "Sessions" button to open the modal, and features hover-based scheduling cues.
- **Features**:
  - Supports `timeGridWeek`, and `dayGridMonth` views with navigation (next, prev, today).
  - Displays three event types: general events, coach availability, and user meetings.
  - Custom event rendering with `EventContent` and styling via `calendarStyles`.
  - Responsive design with fixed 920px height and overflow handling.
- **Props**:
  - `events`, `coachAvailability`, `yourMeetings` for event data.
  - `availableCoachingSessionsData` for draggable sessions (`dragAndDrop`).
  - `onAddEvent` and `onEventDrop` callbacks for event management.
- **Interactivity**:
  - Modal (`ScheduleSession`) for session details (date, time, title).
  - Event handling for drops, clicks, and rescheduling with proper state management.
- **Storybook**: Added stories for both variants, including empty and interactive scenarios.

ObitoDevv commented:
@alebg Hi sir, Can you please review this pr ? Please let me know if there is anything you need to modify ?

alebg commented on packages/ui-kit/lib/components/calendar/schedule-session.tsx:
```diff
@@ -0,0 +1,151 @@
+  groupName?: string;
+  sessionName?: string;
+  onClickDiscard?: () => void;
+  onClickSendRequest?: () => void;
```
@ObitoDevv this callback should be passed the inputs from the user

alebg reviewed (changes requested):
@ObitoDevv the story is broken for the calendar

![image](https://github.com/user-attachments/assets/590aa404-c916-4c88-b429-65fde74bb5d4)

- Apart from that, i left an inline comment

- Additionally, in general, this is a component more complicated than the others, so please try to be disciplined regarding the typing and the props. I see you have everywhere optional props. In particular I don't see why the callbacks should be optional. Then, some things like "Event" could easily be typed as a union type of coach availability and coaching sessions, which would make using your component easier. Another example: it's very opaque to have an optional "extendedProps" in the event type. When should we use it? At the very least, please provide a docstring, but it would be even better if you type your structures properly and avoid optionals as much as possible, so just from the typing its clear for any developer what should be passed in this case.

You can discuss this further with @anilrise11 under the light of the discussions we've been having the past two weeks regarding union types and optional props.

ObitoDevv commented:
> @ObitoDevv the story is broken for the calendar
> 
> ![image](https://private-user-images.githubusercontent.com/62880920/440796782-590aa404-c916-4c88-b429-65fde74bb5d4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY1Mzg4NjYsIm5iZiI6MTc0NjUzODU2NiwicGF0aCI6Ii82Mjg4MDkyMC80NDA3OTY3ODItNTkwYWE0MDQtYzkxNi00Yzg4LWI0MjktNjVmZGU3NGJiNWQ0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA2VDEzMzYwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlmMTA1MmQxMGZjYzBhZDgxMWExZmRmMzYxNzVlZTk3Y2IzMzZhYTMwNDZmNDBiYmIzNzU4ODI5MTY3ODIwYWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.sIilxAta13TDAc4crffAjGMS_NquGvjC3s2YSGZ60cI)
> 
> * Apart from that, i left an inline comment
> * Additionally, in general, this is a component more complicated than the others, so please try to be disciplined regarding the typing and the props. I see you have everywhere optional props. In particular I don't see why the callbacks should be optional. Then, some things like "Event" could easily be typed as a union type of coach availability and coaching sessions, which would make using your component easier. Another example: it's very opaque to have an optional "extendedProps" in the event type. When should we use it? At the very least, please provide a docstring, but it would be even better if you type your structures properly and avoid optionals as much as possible, so just from the typing its clear for any developer what should be passed in this case.
> 
> You can discuss this further with @anilrise11 under the light of the discussions we've been having the past two weeks regarding union types and optional props.

@alebg - Sir, Can you please do `pnpm i` before runing the storybook ? It will work fine.

alebg commented:
> @alebg - Sir, Can you please do `pnpm i` before runing the storybook ? It will work fine.

@ObitoDevv I'm already doing

```sh
pnpm cache delete && rm -f pnpm-lock.yaml && pnpm install --force && pnpm nx run-many -t build
pnpm nx run ui-kit:storybook
```

Still doesn't work

ObitoDevv commented:
Hi @alebg sir,

Thanks for your feedback. I've made the following updates to address your concerns:

- **Calendar Story Fix**: Fixed the broken calendar story in Storybook. The issue stemmed from a misconfigured prop in the calendar component, which caused rendering failures. It's now fully functional.
- **Session ID Handling**: In `AvailableCoachingSessions`, replaced the temporary ID generation (`${session.title}-${index}`) with the `session.id` field provided by the backend, as you suggested. The session model now includes the `id` field for proper identification.
- **Prop Typing Improvements**:
  - Removed optional modifiers for callbacks (`onClickDiscard`, `onClickSendRequest`, `onClickCoach`, etc.) in `ScheduleSession` and similar components to enforce their usage.
  - Refactored the `Event` type into a union type (`CalendarEvent | CoachAvailabilityEvent | UserMeetingEvent`) to improve type safety and make the component's event structure clearer.
  - Eliminated the optional `extendedProps` in event types, replacing it with explicit fields (e.g., `coachingSessionId`, `isCoachAvailability`) to avoid ambiguity.
- **Callback Enhancement**: Modified `onClickSendRequest` in `ScheduleSession` to pass user inputs (date, time, session name, etc.) as parameters, per your inline comment, ensuring the callback provides necessary context.
- **Docstrings**: Added comprehensive docstrings to `AvailableCoachingSessions`, `CoachingSessionCalendar`, and `ScheduleSession`, detailing props, usage examples, and variant behaviors to enhance developer clarity.

alebg commented:
Hi @ObitoDevv ! Thanks a lot. I see the CI broke, could you please fix the errors before I review again?

alebg commented:
@ObitoDevv could you please check if replacing line 26 of `packages/ui-kit/.storybook/main.ts` by:

```ts
        include: [
          'storybook-dark-mode',
          '@fullcalendar/react',
          '@fullcalendar/core',
          '@fullcalendar/daygrid',
          '@fullcalendar/timegrid',
          '@fullcalendar/interaction',
          '@fullcalendar/common', 
        ],
```

causes you any trouble? That's one way I found to make it work for me since it was still broken

alebg commented:
@ObitoDevv Carla has the same problem as me. Please use my suggested fix if that doesn't cause any problems on your side

alebg commented:
@carlapbg a general question:

![immagine](https://github.com/user-attachments/assets/577d7d09-ee58-4601-adaf-3854a9bb148a)

So, we may have available coaching sessions of less than an hour. E.g., 20 mins, 30 mins. However, in the figma it more or less looks that you can only book one session per *hour* even if the session lasts less. Or at least there doesn't seem to be enough space to put many of these smaller sessions. Do you have a suggestion on how to handle this?

alebg reviewed (changes requested):
@ObitoDevv thanks a lot, it's looking good. There's one thing we need to discuss with @carlapbg and I'll come back to you. Apart from that:

- You should only be able to drag and drop or click sessions *within* an available coaching session. For example now I can do this: 

![immagine](https://github.com/user-attachments/assets/ff824e17-a12c-48a7-b02f-17f1bf5f8ccf)

But it doesn't make much sense, because the coach has not marked an availability in that block. Something similar happens in the click variant

- In the modal, I think you should be able to modify the hour, and possibly not the date, and only within the bounds of an available coaching session. This may depend however on what we decide with @carlapbg regarding the design

- This button should read (localized) "Schedule session", as per the figma

![immagine](https://github.com/user-attachments/assets/953c2949-412e-440b-901e-daa844f42253)

carlapbg reviewed (changes requested):
Thank you! I have some comments:

- When you're doing drag & drop, the entire option should have a background color (which depends on whether you're selecting an available slot or not)
![Screenshot 2025-05-10 at 15 42 32](https://github.com/user-attachments/assets/04b54ec5-7333-40f0-91a4-2b3ff2f01536)

But this is how the drag&drop looks now in the storybook:
<img width="848" alt="Screenshot 2025-05-10 at 15 37 51" src="https://github.com/user-attachments/assets/7814afd2-0aaa-4e64-8202-b12323377337" />


Now in the monthly view specifically:
- The calendar shouldn't have those white borders
<img width="1144" alt="Screenshot 2025-05-10 at 15 19 52" src="https://github.com/user-attachments/assets/ebe672d6-73c3-43e1-9504-b110cec8afd0" />

- It should only show the days at the top of the calendar, no day numbers
<img width="1149" alt="Screenshot 2025-05-10 at 15 23 11" src="https://github.com/user-attachments/assets/1eea6a32-09d5-4f4f-86f5-08e6521022a4" />

- Today’s date should be highlighted with a circle around the number, as shown in [Figma](https://www.figma.com/design/8KEwRuOoD5IgxTtFAtLlyS/Just_Do_Ad-1.2?node-id=5223-25247&t=c8VKna8SSST01PZJ-4)
![Screenshot 2025-05-10 at 15 25 10](https://github.com/user-attachments/assets/20a5cf33-44b7-417a-8d3b-1ce8e94d1aed)

 - On smaller screens, events are overlapping
<img width="709" alt="Screenshot 2025-05-10 at 15 31 49" src="https://github.com/user-attachments/assets/0ff928c5-b3d8-4cdc-a573-aea2fcbf9470" />

- When there are too many events on a single day, show 2 or 3, then a “+2 more” indicator. Like this example in the figma:
![Screenshot 2025-05-10 at 15 34 59](https://github.com/user-attachments/assets/862b9b69-a8db-4641-8f56-2cc0e116b94c)

ObitoDevv commented:
@alebg I have one question for varient #184 . Should I use this have branch and push the code ? Or I have to create the new pr for it?

alebg commented:
> @alebg Hi sir, I have one question for varient #184 . Should I use this branch and push the code ? Or I have to create the new pr for it?

Hi @ObitoDevv , sorry for the delay. If you're already working on variant 3 (#184) then yes, please use this same PR (push your commits here)

alebg commented:
@ObitoDevv @anilrise11 to discuss tomorrow:

We decided to modify a bit how this will look, regarding multiple less-than-an-hour sessions per hour in the week view. So we'll allow for this, and visually the minimum will be marked by 15 min thresholds. So from 0 to 15 mins an event will occupy one row, from 15 to 30 mins it will occupy two rows, and so on. So in call cases we'll display all information on hover, and in the smaller events we'll prioritize: (1) title; (2) hour; (3) duration. So only events of at least 45 mins will look like in the figma.gg

This, and we'll let the users input any hour in the modal that pops up

alebg commented:
Hi @ObitoDevv , are we supposed to review this? the stories don't really seem to work as intended. CoachingSessionCalendar in particular, the click variation doesnt show on hover the type of session and on click nothing seems to happen. Also, none of the stories have available slots, such that we can test the whole flow. And when hovering over the component, we see this out of place text:

![immagine](https://github.com/user-attachments/assets/0410e1b8-ca95-41c5-a4c8-cc0c421458f8)

ObitoDevv commented:
Hi @alebg sir, thanks for your feedback! In the CoachingSessionCalendar, we have two variants: a drag-and-drop variant and a click-to-schedule variant. The CoachAvailabilityCalendar includes an "Add Availability" button for coaches to set their availability. The issue you're seeing with the CoachingSessionCalendar is due to the mock data in Storybook being set to a week with no available slots, which prevents interaction in both variants. To see the intended functionality, including the session type on hover and click interactions, please navigate to the previous week in the calendar where coach availability is populated. Let me know if you need further clarification or if there’s anything else to address!

alebg commented on packages/ui-kit/lib/components/calendar/coaching-session-calendar.tsx:
```diff
@@ -0,0 +1,389 @@
+              user="student"
+              isError={false}
+              groupSession={true}
+              coachName="John Doe"
```
@ObitoDevv thanks. So I asked if we were supposed to review this PR properly because as soon as I start looking at the code, I find things like this: hard-coded values, non-localized English strings, unused variables, and so forth. Could you please cleanup this PR all around and then let us know when we can give it a proper review?

alebg reviewed (changes requested):
@ObitoDevv thanks, please check my inline comment

ObitoDevv commented:
@alebg sir,
- Added 'today' label to dictionaries, updated Calendar for localized date formatting.
- Added locale support to `DateInput` in `AvailabilityManager` and `ScheduleSession`.
- Refactored `CoachAvailabilityCalendar` and `CoachingSessionCalendar` story titles for consistency.
- Fixed memory leaks by destroying `Draggable` instances in `AvailableCoachingSessions`.
- Current alert logic (For click to schedule varient) [Included in course]  :
  ```jsx
  if (!isWithinAvailability) {
    alert('No coach availability for this time slot');
    return;
  }
  ```
  - **Question**: Should we replace `alert` with `react-toastify` to display a localized, non-blocking toast message for the "click to schedule" variant, ensuring consistency with the localized UI?
- Title variable:
  ```jsx
  const title = 'Coaching Session';
  ```
  - **Question**: Since Figma only specifies displaying date and time, should we pass the `title` to the backend for storage or remove it from the component to align with the description added ? Below u can check. 
![image](https://github.com/user-attachments/assets/8678f9bb-d5ed-4f84-8b31-da9ee7e4cdb1)

ObitoDevv commented:
@alebg @maany sir, 
I have to create new branch for this calendar-old-pr. But I'm facing one issue -

Issue : When trying to install `FullCalendar dependencies` , we hit a blocking error due to a missing package `@dream-aim-deliver/dad-cats` that doesn't exist in the ` npm registry`.

Root Cause: The dependency was referenced in both the root [package.json](vscode-file://vscode-app/c:/Users/Obito/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) and [package.json](vscode-file://vscode-app/c:/Users/Obito/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html), causing pnpm to fail during dependency resolution across the entire workspace.


Solution (Please confirm this) :

Removed problematic dependencies from both:

Root [package.json](vscode-file://vscode-app/c:/Users/Obito/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html): Removed @dream-aim-deliver/dad-cats and server: "link:@trpc/next-layout/server"
[package.json](vscode-file://vscode-app/c:/Users/Obito/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html): Removed @dream-aim-deliver/dad-cats
Installed FullCalendar dependencies successfully:
```
cd packages/ui-kit
pnpm add @fullcalendar/react @fullcalendar/core @fullcalendar/daygrid @fullcalendar/timegrid @fullcalendar/interaction
```

alebg commented:
@ObitoDevv we will take over this task completely, no need to do anything else on your part

@anilrise11 for other tasks, could you please share with @ObitoDevv the instructions + token ?
